packages/core-database/lib/interface.js

```diff
@@ -307,13 +309,49 @@ module.exports = class ConnectionInterface {
     if (nextRound === round + 1 && height >= maxDelegates) {
       logger.info(`Back to previous round: ${round} :back:`)
 
-      this.blocksInCurrentRound = await this.__getBlocksForRound(round)
-      this.activeDelegates = await this.getActiveDelegates(height)
+      this._activeDelegates = await this.__calcPreviousActiveDelegates(round)
 
       await this.deleteRound(nextRound)
     }
   }
 
+  /**
+   * Calculate the active delegates of the previous round. In order to do
+   * so we need to go back to the start of that round. Therefore we create
+   * a temporary wallet manager with all delegates and revert all blocks
+   * and transactions of that round to get the initial vote balances
+   * which are then used to restore the original order.
+   * @param {Number} round
+   */
+  async __calcPreviousActiveDelegates (round) {
```

Comment:
Sounds a lot like copy on write. But since I only copy delegate wallets this is not much of a concern right now unless we suddenly grow exponentially to >100000 delegates... then we should worry and implement what you described.